### PR TITLE
PP-11681 Replace 'base-client' with 'axios-base-client' in 'ledger.client'

### PR DESCRIPTION
**Pull request by JFSGDS**
https://payments-platform.atlassian.net/browse/PP-11681

## WHAT
- Change ‘ledger.client’ methods to use ‘axios-base-client’.
- Refactor call parameters initialisation.
- Use ‘async/await’ syntax.

## HOW 
All automated tests should pass without the need to be amended.
Existing functionality should remain unaffected.

### CODE REVIEW

File: app/services/clients/ledger.client.js [hunk 1]
```diff
@@ -18,53 +20,50 @@ const defaultOptions = {
...
-      account_id: gatewayAccountId,
-      ...options.transaction_type && { transaction_type: options.transaction_type }
-    },
-    description: 'Get individual transaction details',
-    transform: legacyConnectorTransactionParity
-  }, defaultOptions, options)
-  return baseClient.get(configuration)
+const transaction = async function transaction (id, gatewayAccountId, options = {}) {
+  this.client = new Client(defaultOptions.service)
+  const baseUrl = options.baseUrl ? options.baseUrl : defaultOptions.baseUrl
```
Comment (iqbalgds) [inline app/services/clients/ledger.client.js:25]: Can we make this the same as line37?

File: app/services/clients/ledger.client.js [hunk 2]
```diff
@@ -18,53 +20,50 @@ const defaultOptions = {
...
-    url: `/v1/transaction/${id}/transaction`,
-    qs: {
-      gateway_account_id: gatewayAccountId,
-      transaction_type: 'DISPUTE'
-    },
-    description: 'Get disputes for payment'
-  }, defaultOptions, options)
-  return baseClient.get(configuration)
+async function getDisputesForTransaction (id, gatewayAccountId, options = {}) {
+  const baseUrl = options.baseUrl ? options.baseUrl : defaultOptions.baseUrl
```
Comment (iqbalgds) [inline app/services/clients/ledger.client.js:46]: Do we need to add the options.baseUrl?

**Review by iqbalgds — CHANGES_REQUESTED** [review]
Looks really - just some minor comments

File: app/services/clients/ledger.client.js [hunk 3]
```diff
@@ -18,53 +20,50 @@ const defaultOptions = {
...
-  const configuration = Object.assign({
-    url: `/v1/transaction/${id}`,
-    qs: {
-      override_account_id_restriction: true
-    },
-    description: 'Get individual transaction details with no accountId restriction'
-  }, defaultOptions, options)
-  return baseClient.get(configuration)
+const transactionWithAccountOverride = async function transactionWithAccountOverride (id, options = {}) {
+  const url = urlJoin(defaultOptions.baseUrl,'/v1/transaction', id)
```
Comment (iqbalgds) [inline app/services/clients/ledger.client.js:37]: Do we need to add a check as per line 24?
Check if the baseUrl is the default URL / coming from options object.

File: app/services/clients/ledger.client.js (hunk 2 above)
Comment (iqbalgds) [reply app/services/clients/ledger.client.js:46]: We have discussed this - resolving comment.

File: app/services/clients/ledger.client.test.js [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    it('should use configured base url', async () => {
+      const ledgerClient = getLedgerClient()
+
+      await ledgerClient.transaction('id', 'a-gateway-account-id', { baseUrl: 'https://example.com' })
+
+      expect(configureSpy.getCall(0).args[0]).to.equal('https://example.com/v1/transaction/id?account_id=a-gateway-account-id')
+    })
+  })
+
+  describe('transactionWithAccountOverride function', () => {
```
Comment (iqbalgds) [inline app/services/clients/ledger.client.test.js:106]: In the code, this function seems to only use the default base URL option.
So not sure why this is working?
We can chat about this.

File: app/services/clients/ledger.client.test.js [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      expect(configureSpy.getCall(0).args[0]).to.equal('http://127.0.0.1:8006/v1/transaction/id?override_account_id_restriction=true')
+    })
+
+    it('should use configured base url', async () => {
+      const ledgerClient = getLedgerClient()
+
+      await ledgerClient.transactionWithAccountOverride('id', { baseUrl: 'https://example.com' })
+
+      expect(configureSpy.getCall(0).args[0]).to.equal('http://127.0.0.1:8006/v1/transaction/id?override_account_id_restriction=true')
+      // expect(configureSpy.getCall(0).args[0]).to.equal('https://example.com/v1/transaction/id?override_account_id_restriction=true')
```
Comment (iqbalgds) [inline app/services/clients/ledger.client.test.js:125]: Commented out test.

Comment (JFSGDS) [reply app/services/clients/ledger.client.test.js]: Uncommented as suggested.

File: app/services/clients/ledger.client.test.js (hunk 4 above)
Comment (JFSGDS) [reply app/services/clients/ledger.client.test.js:106]: Fixed by adding check to to see if options parameter contains baseUrl and using that if present.

File: app/services/clients/ledger.client.js (hunk 1 above)
Comment (iqbalgds) [reply app/services/clients/ledger.client.js:25]: This line is correct - resolving comment

**Review by iqbalgds — DISMISSED** [review]
Good work - would be good to get another set of eyes to check one of these PRs.

**Review by iqbalgds — APPROVED** [review]
LGTM - just need to squash commits.

**Pull request merged**